arangod/RocksDBEngine/RocksDBTransactionState.cpp

```diff
@@ -152,113 +144,63 @@ Result RocksDBTransactionState::beginTransaction(transaction::Hints hints) {
   auto& selector = vocbase().server().getFeature<EngineSelectorFeature>();
   auto& engine = selector.engine<RocksDBEngine>();
   rocksdb::TransactionDB* db = engine.db();
-  _rocksReadOptions.prefix_same_as_start = true;  // should always be true
 
-  _rocksReadOptions.fill_cache = options().fillBlockCache; 
-
-  TRI_ASSERT(_readSnapshot == nullptr);
   if (isReadOnlyTransaction()) {
-    // no need to acquire a snapshot for a single op
-    if (!isSingleOperation()) {
-      _readSnapshot = db->GetSnapshot();  // must call ReleaseSnapshot later
-      TRI_ASSERT(_readSnapshot != nullptr);
-      _rocksReadOptions.snapshot = _readSnapshot;
+    if (isSingleOperation()) {
+      _rocksMethods = std::make_unique<RocksDBSingleOperationReadOnlyMethods>(this, db);
+    } else {
+      _rocksMethods = std::make_unique<RocksDBReadOnlyMethods>(this, db);
     }
-    _rocksMethods = std::make_unique<RocksDBReadOnlyMethods>(this);
   } else {
-    createTransaction();
-    TRI_ASSERT(_rocksTransaction != nullptr);
-
-    _rocksReadOptions.snapshot = _rocksTransaction->GetSnapshot();
-    if (hasHint(transaction::Hints::Hint::INTERMEDIATE_COMMITS)) {
-      TRI_ASSERT(_options.intermediateCommitCount != UINT64_MAX ||
-                 _options.intermediateCommitSize != UINT64_MAX);
-      _readSnapshot = db->GetSnapshot();  // must call ReleaseSnapshot later
-      TRI_ASSERT(_readSnapshot != nullptr);
+    if (isSingleOperation()) {
+      _rocksMethods = std::make_unique<RocksDBSingleOperationTrxMethods>(this, db);
+    } else {
+      _rocksMethods = std::make_unique<RocksDBTrxMethods>(this, db);
     }
+  }
+  _rocksMethods->beginTransaction();
```

Comment:
AFAIK `beginTransaction` returns a `Result`, so we should check here if it has failed or not.